metadata/src/main/java/org/apache/kafka/image/FeaturesDelta.java

```diff
@@ -58,7 +59,14 @@ public void finishSnapshot() {
 
     public void replay(FeatureLevelRecord record) {
         if (record.name().equals(MetadataVersion.FEATURE_NAME)) {
-            metadataVersionChange = MetadataVersion.fromFeatureLevel(record.featureLevel());
+            // Support for the `metadata.version` feature flag was added in IBP_3_3_IV0, so it's possible (but unlikely) that we read
+            // records with a feature level that is no longer supported for clusters that used a pre-release version of 3.3.0.
+            // We automatically fallback to IBP_3_3_IV3 in that case. We use explicit versions instead of `MINIMUM_VERSION` because
+            // we want to force an explicit decision if we change `MetadataVersion.MINIMUM_VERSION` in the future.
+            if (record.featureLevel() >= MINIMUM_PERSISTED_FEATURE_LEVEL && record.featureLevel() <= MetadataVersion.IBP_3_3_IV3.featureLevel())
+                metadataVersionChange = MetadataVersion.IBP_3_3_IV3;
```

Comment:
@junrao @cmccabe Thoughts on this fallback? It seems like quite a rare edge case (since it requires usage of a pre-release version and for the older metadata version not to be compacted away if I understand correctly), but it seems fair to never fail while reading the log and the fallback code is pretty simple.

Reply:
Is it guaranteed that we won't have a record with the older metadata version in the log in that case?

Reply:
There is no guarantee, but the expected MV is IBP_3_3_IV3 and above. If this is not the case, it seems easier to just fail instead of changing it silently.

Reply:
I discussed this with a few people and the bit that was missing in the discussion above is that we generate snapshots every hour by default and hence this issue is unlikely to ever happen and the workaround is reasonably simple. I updated the code to remove the fallback and include a helpful message if this issue occurs.